app/src/main/java/com/group80/uoftinder/ContactsActivity.java

```diff
@@ -0,0 +1,134 @@
+            @Override
+            public ContactViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+                View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.uoftinder_contact_list_contact_layout, parent, false);
+                return new ContactViewHolder(view);
+            }
+        };
+
+        RecyclerView recyclerView = findViewById(R.id.contactActivityRecyclerView);
+        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
+        recyclerView.setHasFixedSize(true);
+        recyclerView.setLayoutManager(linearLayoutManager);
+        recyclerView.setAdapter(contactAdapter);
+    }
+
+    @Override
+    public void onStart() {
+        super.onStart();
+        contactAdapter.startListening();
+    }
+
+    @Override
+    public void onStop() {
+        super.onStop();
+        if (contactAdapter != null) {
+            contactAdapter.stopListening();
+        }
+    }
+
+    private class ContactViewHolder extends RecyclerView.ViewHolder {
+        protected final TextView contactName;
+        protected final ImageView contactPic;
+
+        public ContactViewHolder(@NonNull View itemView) {
+            super(itemView);
+            contactName = itemView.findViewById(R.id.contactListContactName);
+            contactPic = itemView.findViewById(R.id.contactListContactPic);
+        }
+
+        /**
+         * Triggered when this View Holder is bind to an adapter
+         * <p>
+         * // TODO: this method may require architectural fix
+         *
+         * @param context      the global information about an application environment // TODO: architectural fix?
+         * @param position     the position of this View Holder in the adapter
+         * @param contactModel a model storing the contact information // TODO: architectural fix
+         */
+        protected void onBind(Context context, int position, @NonNull ContactModel contactModel) {
+            contactName.setText(contactModel.getName());
+
+            // TODO: extract the set-to-image-view process?
+            ImageStorageDbFacade.downloadImage(
+                    new String[]{contactModel.getUid(), "img", "_profile_img.jpg"},
+                    new StorageDbDownloadable<byte[]>() {
+                        @Override
+                        public void onStorageDownloadSuccess(byte[] data) {
+                            contactPic.setImageBitmap(BitmapFactory.decodeByteArray(data, 0, data.length));
+                        }
+
+                        @Override
+                        public void onStorageDownloadFailure(@NonNull Exception exception) {
+                            // Set to a default profile image
+                            contactPic.setImageDrawable(AppCompatResources.getDrawable(getApplicationContext(), R.drawable.ic_baseline_account_circle_24));
+                        }
+                    }
+            );
+
+            // TODO: optimize architecture
```

Comment:
I feel like this code seems like something that the view should not be doing, as u mentioned. This seems more like a Use Case and should be passed into a Controller that passes the necessary parameters to the Use Case

Reply:
Yes, I agree with Nick here. Seems like you're handling a lot of the logic required for this use case directly in the view.

Reply:
I do agree the architecture is not so clean here. I am working on migrating to MVP and hopefully things will be cleaner after the transformation has been done.